File: B트랙/7주차/참고 코드/back_tracking.py
# https://www.acmicpc.net/problem/15649
import sys
input = sys.stdin.readline

def dfs(depth, num, limit):
    if depth == limit:
        print(' '.join(map(str,ary)))
        return
    for i in range(1, num+1):
        if visited[i]:
            continue
        visited[i] = True
        ary.append(i)
        dfs(depth+1,num,limit)
        ary.pop()
        visited[i] = False

n,m = map(int,input().strip().split())
sys.setrecursionlimit(m + 10)
visited = [False for _ in range(n+1)]
ary = []
dfs(0,n,m)

File: B트랙/7주차/참고 코드/back_tracking.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>

void dfs(int* visited, int* ary, int depth, int num, int limit)
{
	if (depth == limit)
	{
		for (int i = 0; i < depth; ++i)
		{
			printf("%d ", ary[i]);
		}
		printf("\n");
		return;
	}
	for (int i = 1; i <= num; ++i)
	{
		if (visited[i])
		{
			continue;
		}
		visited[i] = 1;
		ary[depth] = i;
		dfs(visited, ary, depth + 1, num, limit);
		ary[depth] = 0;
		visited[i] = 0;
	}
}

int main(void) // https://www.acmicpc.net/problem/15649
{
	int n = 0, m = 0;
	scanf("%d %d", &n, &m);
	int* visited = (int*)calloc(n + 1, sizeof(int));
	int* ary = (int*)calloc(m + 1, sizeof(int));
	dfs(visited, ary, 0, n, m);
	free(visited);
	free(ary);
	return 0;
}